### DISCUSSION TITLE
Wheel names for 1.4 are so long that they break poetry

### DISCUSSION THREAD

**Discussion by andyhasit** [sqlalchemy/sqlalchemy#10600] [Usage Questions]
None of the categories in the issue tracker seemed to cater for this so I thought I'd put it here.

The wheel names for 1.4 are so long that they break poetry, as reported [here](
https://github.com/python-poetry/poetry/issues/8529).

It's not uncommon for packages to have rather long names when concatenating multiple targets, but SQLAlchemy 1.4 pushes the boat, which causes poetry to error during installation, and the workarounds are painful.

Neither PIP nor PYPI set  limits on filename, but that doesn't mean we should go crazy. SQLAlchemy 2.0 seems to have brought it back to a reasonable length so I'm presuming we can produce a patch version for 1.4 which fixes this?

```
greenlet-3.0.1-cp310-cp310-manylinux_2_24_x86_64.manylinux_2_28_x86_64.whl
requests-2.31.0-py3-none-any.whl
urllib3-2.0.7-py3-none-any.whl
SQLAlchemy-1.4.49-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl
SQLAlchemy-1.4.50-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl
SQLAlchemy-2.0.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl

```

**Comment by CaselIT** [comment]
Hi,

sqlalchemy doesn't decide the name of the wheel, it's named automatically by the build system based on the wheel spec by auditwheel. It'd done like this to support ancient versions of linux.

honestly I don't understand the issue, you seem to be using linux that has no stupid file name length restriction like windows (that has workarounds but it's not of interest here), so why is this a problem?

I guess use pip is not an acceptable solution?

**Reply by CaselIT** [reply]
TIL that linux has 255 bytes file name limit, but that does not explain the issue. that file name is not 255 bytes, so again why is this an issue? Is poetry writing files with names in utf-32?

**Reply by zzzeek** [reply]
what makes our wheel names "support ancient versions of linux" and is not the case with other packages?  do they "not support ancient versions of linux" or do they have separate files for those ancient versions ?

**Reply by CaselIT** [reply]
> I guess use pip is not an acceptable solution?

more realistic workaround, use the no binary option to install the source python file. It will get compiled if possible, else install as pure python version.

**Reply by CaselIT** [reply]
> what makes our wheel names "support ancient versions of linux" and is not the case with other packages? do they "not support ancient versions of linux" or do they have separate files for those ancient versions ?

I think it's because we are using also manylinux1, that was a thing when we setup 1.4 builds. in v2 where we use cibuildwheel that was deprecated (I think). I honestly don't know, wheel for linux are (were) a mess.
Also at some point something changed in the wheel specification that created that long names, they where not like that.
Example 
- https://pypi.org/project/SQLAlchemy/1.4.13/#files - old version with short files
- https://pypi.org/project/SQLAlchemy/1.4.14/#files - new version with long files

something changed on may 2021

**Reply by CaselIT** [reply]
the image used to build the file changed, using an updated version of auditwheel (I guess, haven't checked).
https://github.com/sqlalchemy/sqlalchemy/compare/rel_1_4_13...rel_1_4_14

all this is managed by python packaging people, so if they are fine with long names then I think the bug is on poetry side

**Comment by zzzeek** [comment]
these names have nothing to do with SQLAlchemy.   Shouldn't this be an issue in Poetry?  It's quite unusual in 2023 for "too long filename" to be a thing for a Python library?

**Reply by CaselIT** [reply]
> Everything old is new again

**Reply by andyhasit** [reply]
I raised it with poetry first. The issue is saving the wheel to their cache, which has a long path already, and then hitting the OS path limit when trying to do anything with that file. See the issue linked in first message.

We're upgrading from 1.3 to 1.4 and it didn't seem to be an issue with 1.3 (for which I can't seem to download wheels directly to check the file length right now) not do I see it being a problem with 2.0, but we need to go via 1.4.

I'm surprised this is a new issue, surely a ton of people are installing 1.4 using poetry...

As for just using pip - that is the work around but it's clunky when your whole CI and deployment runs off poetry.

If SQLalchemy can't change the name and poetry won't fix then another solution is to move poetry's cache dir closer to root.

**Reply by CaselIT** [reply]
> then hitting the OS path limit

this is strange. tried looking and every source says that for linux there is 255 bytes limit of file name and 4096 bytes for path limitation. These seem to be sufficient here for doing any path manipulation (the file name is 115 character long)
So maybe it's something that's configured in a strange way in your system?

> As for just using pip - that is the work around but it's clunky when your whole CI and deployment runs off poetry.

using the no binary option may also be an option

> I'm surprised this is a new issue, surely a ton of people are installing 1.4 using poetry...

honestly this was never raised. It's also the fist time I see such an issue on linux.
